File: scripts/test-support/python_coverage.py
"""coverage.py instrumentation for the Pyodide-hosted webchirp_bridge package.

Executed into Pyodide's globals by scripts/test-support/python-coverage.mjs.
The runtime namespace is flat (web/python/runtime_bridge.py flattens the
package into it), so these names sit alongside the bridge's own and are
prefixed to keep them out of its way.

Deliberately not under web/python/: nothing here ships to the browser. It
exists so npm run coverage can measure the Python half of the runtime, and it
is seeded only when scripts/test-radio-harness.mjs is asked for coverage.

web/python/runtime_bridge.py itself is not measurable. It is the one runtime
file that is executed as a string rather than imported, so CPython compiles it
under the name "<exec>" and coverage cannot map it to a path. The merge step
reports it as uninstrumented rather than silently dropping it.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    # Annotation-only, and gated so it is not imported at runtime: coverage is
    # loaded into Pyodide by _webchirp_coverage_start(), so importing it at
    # module scope would fail on the seed that happens before loadPackage().
    import coverage

# Where seedPyodideRuntime() writes the package inside Pyodide's filesystem.
_PACKAGE_DIR = "/webchirp_runtime/webchirp_bridge"

# The live Coverage object between start and fragment, or None outside that
# window. Named concretely rather than as Any so a reader knows what it holds.
_coverage: coverage.Coverage | None = None


def _webchirp_coverage_start() -> None:
    """Begin tracing every module of the bridge package.

    Uses include= rather than source=. Coverage has to start before
    seedPyodideRuntime() writes the package, and source= resolves a path that
    does not exist yet as a module name instead, then reports "No data was
    collected" -- a silent zero rather than an error.
    """
    global _coverage
    import coverage

    _coverage = coverage.Coverage(data_file=None, include=[f"{_PACKAGE_DIR}/*"])
    _coverage.start()


def _webchirp_coverage_fragment() -> str:
    """Stop tracing and return this process's slice of the result as JSON.

    Every .py in the package is analysed, not only the ones that were imported,
    so a module this process never loaded is reported with all of its
    statements missing rather than omitted. Otherwise Python coverage would
    have the same blind spot V8 coverage has, where an unloaded file is absent
    from the denominator instead of scoring zero.

    Paths are Pyodide filesystem paths; the caller maps them back to
    web/python/. Line numbers are absolute and need no adjustment.
    """
    if _coverage is None:
        return json.dumps({})
    _coverage.stop()
    files: dict[str, dict[str, list[int]]] = {}
    for path in sorted(Path(_PACKAGE_DIR).glob("*.py")):
        name = str(path)
        # analysis2() parses the file whether or not it was ever imported, and
        # returns (filename, statements, excluded, missing, missing_formatted).
        _, statements, _excluded, missing, _formatted = _coverage.analysis2(name)
        executed = sorted(set(statements) - set(missing))
        files[name] = {"statements": sorted(statements), "executed": executed}
    return json.dumps(files)

File: scripts/test-support/python-coverage.mjs
// Env-gated coverage.py instrumentation for the Pyodide runtime, wired into
// scripts/test-radio-harness.mjs. Off unless WEBCHIRP_PY_COVERAGE names a
// directory, so an ordinary npm test pays nothing: no extra package load, no
// tracer, no exit hook.
//
// Pyodide ships coverage.py in its package set and its C tracer works under
// WASM, so this measures the real thing rather than a re-implementation. The
// tracer itself is Python and lives in scripts/test-support/python_coverage.py;
// this module only seeds it and moves data across the JS boundary.
//
// node:test gives each test file its own process, and each process boots its
// own Pyodide, so every one writes a fragment and scripts/coverage.mjs merges
// them. Merging is a union of executed lines over a union of statements: a
// module one process never imported still carries its full statement list from
// that process's fragment, so nothing drops out of the denominator.
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const supportDir = path.dirname(fileURLToPath(import.meta.url));
const TRACER_PYTHON_PATH = path.join(supportDir, "python_coverage.py");

// Where seedPyodideRuntime() writes the runtime, and where those files live in
// the repo. Fragments carry Pyodide paths; callers want repo-relative ones.
const RUNTIME_PREFIX = "/webchirp_runtime/";
const REPO_PREFIX = "web/python/";

export const PYTHON_COVERAGE_ENV = "WEBCHIRP_PY_COVERAGE";

// The directory fragments are written to, or "" when coverage is off.
export function pythonCoverageDir() {
  const configured = String(process.env[PYTHON_COVERAGE_ENV] || "").trim();
  return configured ? path.resolve(configured) : "";
}

// Turn a Pyodide filesystem path into the repo path the same file has on disk,
// so a merged report and an lcov file both point at something a reader (or
// GitHub's diff view) can open.
export function toRepoPath(runtimePath) {
  const text = String(runtimePath || "");
  return text.startsWith(RUNTIME_PREFIX) ? REPO_PREFIX + text.slice(RUNTIME_PREFIX.length) : text;
}

// Start tracing in a freshly loaded Pyodide, before the runtime is seeded --
// tracing has to be running while webchirp_bridge is imported or the
// module-level statements never register as executed. Returns whether it did
// anything, so the caller can stay quiet when coverage is off.
export async function startPythonCoverage(pyodide) {
  const outputDir = pythonCoverageDir();
  if (!outputDir) {
    return false;
  }
  await pyodide.loadPackage("coverage");
  pyodide.runPython(fs.readFileSync(TRACER_PYTHON_PATH, "utf8"));
  pyodide.runPython("_webchirp_coverage_start()");

  // process.on("exit") only runs synchronous work, which is all this needs:
  // runPython and writeFileSync are both sync. An async hook would be dropped.
  let written = false;
  process.on("exit", () => {
    if (written) {
      return;
    }
    written = true;
    try {
      writePythonCoverageFragment(pyodide, outputDir);
    } catch (error) {
      // A missing fragment understates coverage but must never fail a test
      // run that otherwise passed, so this reports rather than throws.
      process.stderr.write(`python coverage fragment failed: ${error && error.message}\n`);
    }
  });
  return true;
}

// Pull this process's coverage out of Pyodide and write it as one fragment
// file. Named by pid plus a counter because a single process can boot more
// than one isolated harness.
let fragmentSeq = 0;
export function writePythonCoverageFragment(pyodide, outputDir) {
  const json = pyodide.runPython("_webchirp_coverage_fragment()");
  const parsed = JSON.parse(json);
  if (Object.keys(parsed).length === 0) {
    return "";
  }
  fs.mkdirSync(outputDir, { recursive: true });
  const target = path.join(outputDir, `fragment-${process.pid}-${fragmentSeq++}.json`);
  fs.writeFileSync(target, JSON.stringify(parsed));
  return target;
}
